### PR TITLE
Support nested macro definitions

### PR DESCRIPTION
**Pull request by Jermolene**
### Introduction

It is common for macros to require further "sub-macros" for their operation. For example, here is an outline of a macro that renders a special button that uses a further macro to contain the action string:

```
\define actions()
<$action-sendmessage ... />
\end

\define special-button()
<$button actions=<<actions>>>
Caption
</$button>
\end
```

That should work fine in most situations, but a particularly pernicious case arises when the macros are made global by being tagged `$:/tags/Macro`. The problem then is that not only does `special-button` get exposed as a global variable, so does the `actions` macro. That means that the `<<special-button>>` macro would not work correctly if it was invoked in a context where the variable `actions` has been redefined. For example:

### Proposal

It is proposed to extend the macro definition syntax to allow the macro name to be specified after the `\end` keyword. When parsing a macro, the end marker is taken to be whichever is found first of an empty `\end` marker or a matching `\end <macroname>` marker.

This makes it possible to nest macros inside one another, completely hiding the nested definitions from the outside world.

The example above could instead be written as:

```
\define special-button()
\define actions()
<$action-sendmessage ... />
\end actions
<$button actions=<<actions>>>
Caption
</$button>
\end special-button
```

Note that the following example does not nest as expected because when parsing the parent definition it will find the first `\end` marker and use that as the terminator:

```
\define parent ( )
 commands, comments, ...
 \define child ()
    ....
 \end
 ....
\end
```

In other words, when nesting macro definitions, only the innermost macro definition can omit the macro name from the `\end` marker.

### To do

- [x] Documentation

### CODE REVIEW

**Comment by kookma** [comment]
This is a good PR and has use cases as you said, or when one wants to encapsulate an object (properties, and methods) in a single macro with several nested macros.

As an example, in Matlab we have three types of function

* Main Function (the first function in a file and the name of file is the same as function, nothing appears before it except comments), visible to all other file and also working space
* Sub function (any function defined after main function) and is visible to all other function in the same file
* Nested function (any function inside other function

Matlab and Fortran and Python use a similar approach like

```
function parent ( )
 commands, comments, ...
 function child ()
    ....
 end 
 ....
end
```


This can be used in any number of nesting levels. Fortran, and Matlab also allow to add the name of function to `end` statement. This is for readability like

```
function parent ( )
 commands, comments, ...
 function child ()
    ....
 end child
 ....
end parent
```


I may suggest a similar approach for nesting

```
\define parent ( )
 commands, comments, ...
 \define child ()
    ....
 \end child
 ....
\end parent
```

Or simply 

```
\define parent ( )
 commands, comments, ...
 \define child ()
    ....
 \end
 ....
\end
```

If parser see `\define name()` in a line then looks for a match `\end` else raises an error

**Comment by Jermolene** [comment]
Thank you @kookma using `\end <macroname>` is a 100x better than multiplying the backslashes. I'll amend this PR as soon as I get a chance.

**Comment by kookma** [comment]
I would be happy to create some examples for docs, ones it is ready to go.

In 5.2.3 and older TW, I have used nested macro, as single line short macros have no requirement for end, so the below code works in 5.2.3

```
\define parent(name:"")
\define child() $name$ middle name is Jack
Hi $name$.
<$list filter="[<__name__>match[Motovun]]"><<child>></$list>
\end

<<parent Micky>>

<<parent "Motovun">>
```
This produces

```

Hi Micky.

Hi Motovun. Motovun middle name is Jack

```

**Comment by saqimtiaz** [comment]
There is an interesting issue that has come up on the discussion forum related to nested macros within action strings: https://talk.tiddlywiki.org/t/a-macro-trap-in-case-of-side-effect-involved/4885

In short, the action strings within nested macro definitions are also executed. Considering the widget tree created one understands why this happens. The question is whether we should try to introduce some mitigation against this while adding better support for nested macros, or just try to document the behaviour?

**Comment by Jermolene** [comment]
Hi @saqimtiaz 

> In short, the action strings within nested macro definitions are also executed. Considering the widget tree created one understands why this happens. The question is whether we should try to introduce some mitigation against this while adding better support for nested macros, or just try to document the behaviour?

I couldn't follow the issue in that thread. Could you give an example of the problem? Is it just the unexpected results of using a macro containing nested macro definitions as the value of an action attribute?

**Comment by Jermolene** [comment]
I've updated the PR as per @kookma's [suggestion above](https://github.com/Jermolene/TiddlyWiki5/pull/7004#issuecomment-1286429236). I also added a note to explain why one of the examples given by @kookma still won't work quite as expected.

**Comment by kookma** [comment]
I will do some experiments and if there is any issue I will share here!

**Comment by Jermolene** [comment]
Thanks @kookma much appreciated.

**Comment by kookma** [comment]
Hi @Jermolene 
I did several tests, and all passed except one!


The one fails

Case 1: No Nesting (code works)

```
\define new-tiddler-text() 
Go to code: <$link to="$(createTiddler-title)$" />
\end new-tiddler-text

\define create-actions()
<$action-createtiddler $basetitle=<<__tidname__>> tags="code" text=<<now>> >
  <$action-createtiddler $basetitle={{{ [<createTiddler-title>addsuffix[/remark]] }}} tags="remark" text=<<new-tiddler-text>> />
</$action-createtiddler>
\end create-actions


\define new-code(tidname, label)
<!-- main macro -->
<$button actions=<<create-actions>> >$label$</$button>

\end new-code

<<new-code tidname:"test001" label:"Make it">>
```

Case ii: Nesting (code fails). The one failes

```
\define new-code(tidname, label)
<!-- main macro -->

\define new-tiddler-text() 
Go to code: <$link to="$(createTiddler-title)$" />
\end new-tiddler-text

\define create-actions()
<$action-createtiddler $basetitle=<<__tidname__>> tags="code" text=<<now>> >
  <$action-createtiddler $basetitle={{{ [<createTiddler-title>addsuffix[/remark]] }}} tags="remark" text=<<new-tiddler-text>> />
</$action-createtiddler>
\end create-actions


<$button actions=<<create-actions>> >$label$</$button>

\end new-code

<<new-code tidname:"test001" label:"Make it">>
```

This line fails `Go to code: <$link to="$(createTiddler-title)$" />` and substitution does not work, an empty value inserted.

**Comment by kookma** [comment]
Further update:

Consider a parent with two children.
The second child calls first one, where the first access to a vars from second child.
It seemes vars like `$()$`

**Comment by pmario** [comment]
>Case ii: Nesting (code fails). The one failes

@kookma ... If something fails please create a new issue as you did for other issues. This makes it sure, that we don't miss it in the "wall of text" we already have with this PR.

**Comment by kookma** [comment]
> > Case ii: Nesting (code fails). The one failes
> 
> @kookma ... If something fails please create a new issue as you did for other issues. This makes it sure, that we don't miss it in the "wall of text" we already have with this PR.

Sure. 

See  https://github.com/Jermolene/TiddlyWiki5/issues/7006

File: editions/tw5.com/tiddlers/wikitext/Macro Definitions in WikiText.tid [hunk 1]
```diff
@@ -17,6 +19,20 @@ Alternatively, the entire definition can be presented on a single line without a
...
+<$action-sendmessage $message="tm-notify" $param="HelloThere"/>
+\end actions
+<$button actions=<<actions>>>
+$caption$
+</$button>
+\end special-button
+
+<<special-button>>
+""">>
+
```
Comment (kookma) [inline editions/tw5.com/tiddlers/wikitext/Macro Definitions in WikiText.tid:35]: Question 1: Is this a requirement to have all nested macros right after `\define parent()`?
Of course, the current logic is fine and makes codes simpler to follow.

Question ii: Is it possible to allow one tab, or few whitespaces before `\define` and `\end` for nested macros?
I understood pragma `\define` shall start from first column!

Comment (Jermolene) [reply editions/tw5.com/tiddlers/wikitext/Macro Definitions in WikiText.tid:35]: > Question 1: Is this a requirement to have all nested macros right after `\define parent()`? Of course, the current logic is fine and makes codes simpler to follow.

Yes – the contents of the parent macro forms a new parsing context that behaves just the same as when the entire body of a tiddler is parsed: pragmas still need to be at the top.

> Question ii: Is it possible to allow one tab, or few whitespaces before `\define` and `\end` for nested macros? I understood pragma `\define` shall start from first column!

I wondered about that, I guess it will be a popular request. I think we'd apply that to all pragmas if we were to change it.

**Comment by kookma** [comment]
Hi @Jermolene 
I put some comments on nested macro docs. Seems I should comment here!  Sorry for inconveniences!

* https://github.com/FlashSystems/TiddlyWiki5/commit/bcb3b06d60a3b616083e9df36fc487cbbbf8abb7#commitcomment-91922651
* https://github.com/FlashSystems/TiddlyWiki5/commit/bcb3b06d60a3b616083e9df36fc487cbbbf8abb7#commitcomment-91922974

**Comment by Jermolene** [comment]
Thanks @kookma, I've made some updates in 28c1e6bfc3b4a842323ab6cc03c34c17a93e45e9
